sdk/core/Azure.Core.Experimental/samples/Variant.md

```diff
@@ -0,0 +1,135 @@
+
+`Azure.Variant` is an implementation of a [tagged union](https://en.wikipedia.org/wiki/Tagged_union).  It can hold both reference and value types and can be used to avoid boxing .NET primitives.  The list of value types that Variant can hold without boxing is below.
+
+<details>
+<summary>Value types that `Variant` won't box</summary>
+
+- `byte`
+- `byte?`
+- `sbyte`
+- `sbyte?`
+- `bool`
+- `bool?`
+- `char`
+- `char?`
+- `short`
+- `short?`
+- `int`
+- `int?`
+- `long`
+- `long?`
+- `ushort`
+- `ushort?`
+- `uint`
+- `uint?`
+- `ulong`
+- `ulong?`
+- `float`
+- `float?`
+- `double`
+- `double?`
+- `DateTimeOffset`
+- `DateTimeOffset?`
+- `DateTime`
+- `DateTime?`
+- Enums
+
+</details>
+
+In principle, `Variant` is similar to `object`, but without boxing as described above.  However, since it's not `object`, there are different APIs needed to achieve some of the same functionality, including:
+
+- [Assigning a value to `Variant`](#assign-a-value-to-variant)
+- [Retrieving the value a `Variant` holds](#get-the-value-from-variant)
+- [Working with `null` and `Variant`](#handling-nulls)
+
+## Assign a value to Variant
+
+`Variant` has implicit cast operators for each of the primitives listed above.  This means you can assign a value of those types to `Variant` without casting, and a new `Variant` instance holding an unboxed copy of the value will be created.
+
+```csharp
+Variant v = 3L;
+
+// v.Type is System.Int64
+```
+
+Since you can't have an implicit cast operator from `object`, for reference types, you must create a new instance of `Variant` and pass the value to the `Variant` constructor.
+
+```csharp
+MemoryStream s = new();
+Variant v = new(s);
+
+// v.Type is System.IO.MemoryStream
+```
+
+`Variant` will box value types that aren't on the list above, because it only stores 16 bytes and user-defined value types might exceed this size.  The one exception is for `enum`, which `Variant` can hold without boxing.  Because we can't add implicit cast operators for unknown enum types, you must call the `Create` method to create a `Variant` that holds an enum without boxing.
+
+```csharp
+Value v = Value.Create(System.Color.Blue);
+
+// v.Type is System.Color
+```
+
+## Get the value from Variant
+
+`Variant` has explicit cast operators for each of the primitives listed above, as well as to `string`.  That means you can assign a `Variant` holding one of these types to a variable of that type with a cast:
+
+```csharp
+Variant v = 3;
+int i = (int)v;
+```
+
+If you don't know the type of the value that a `Variant` is holding, you can read that from the `Variant.Type` property.
+
+```csharp
+switch (v.Type)
+{
+    case Type s when s == typeof(string):
+        Console.WriteLine($"string: {v}");
+        break;
+    case Type i when i == typeof(int):
+        Console.WriteLine($"int: {(int)v}");
+        break;
+}
+```
+
+If you do know the type the `Variant` is holding, you can use the `As<T>` method to get its value as that type.
```

Comment:
As a potential caller, I'm not clear on whether the explicit cast operation, `As<T>`, or both will properly downcast or upcast e.g., `Variant` holds an `int` but would either approach upcast to a `long`? Or maybe it holds a `long` but small enough to fit into an `int`: will that be supported? What happens on overflow? I'd love to see these details explained, even if briefly, to know how to better use `Variant`.

Reply:
Oh. You do explain that `As<T>` doesn't cast just below. But, still, my question stands about the explicit cast operator. And should we support that if not supported already? I see a lot of advantage in some scenarios rather than a potentially long switch statement like you show above.

Reply:
Good questions - let me clarify this!